hs_core/views/resource_folder_hierarchy.py

```diff
@@ -1,456 +1,682 @@
+        return HttpResponse('Bad request - src_path or tgt_path cannot be empty',
+                            status=status.HTTP_400_BAD_REQUEST)
+
+    src_path = str(src_path).strip()
+    tgt_path = str(tgt_path).strip()
+    src_folder, src_base = os.path.split(src_path)
+    tgt_folder, tgt_base = os.path.split(tgt_path)
+    if src_folder != tgt_folder:
+        return HttpResponse('Bad request -- src_path and tgt_path must be in same folder',
+                            status=status.HTTP_400_BAD_REQUEST)
+
+    if not src_path.startswith('data/contents/'):
+        return HttpResponse('Bad request - src_path must start with data/contents/',
+                            status=status.HTTP_400_BAD_REQUEST)
+    if src_path.find('/../') >= 0 or src_path.endswith('/..'):
+        return HttpResponse('Bad request - src_path cannot contain /../',
+                            status=status.HTTP_400_BAD_REQUEST)
+
+    if not tgt_path.startswith('data/contents/'):
+        return HttpResponse('Bad request - tgt_path must start with data/contents/',
+                            status=status.HTTP_400_BAD_REQUEST)
+
+    if tgt_path.find('/../') >= 0 or tgt_path.endswith('/..'):
+        return HttpResponse('Bad request - tgt_path cannot contain /../',
+                            status=status.HTTP_400_BAD_REQUEST)
+
+    istorage = resource.get_irods_storage()
+
+    # check that the target doesn't exist
+    tgt_folder, tgt_base = os.path.split(tgt_path)
+    tgt_folder = tgt_folder[len('data/contents/'):]
+    tgt_abspath = get_path(resource, tgt_base, folder=tgt_folder)
+    if istorage.exists(tgt_abspath):
+        return HttpResponse('Bad request - tgt_path already exists',
+                            status=status.HTTP_400_BAD_REQUEST)
+
+    # check that the source exists
+    src_folder, src_base = os.path.split(src_path)
+    src_folder = src_folder[len('data/contents/'):]
+    src_abspath = get_path(resource, src_base, folder=src_folder)
+    if istorage.exists(src_abspath):
+        return HttpResponse("Bad request - src_path does not exist",
+                            status=status.HTTP_400_BAD_REQUEST)
+
+    try:
+        rename_file_or_folder(user, pk, src_path, tgt_path)
+    except SessionException as ex:
+        return HttpResponse(ex.stderr, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+    except DRF_ValidationError as ex:
+        return HttpResponse(ex.detail, status=status.HTTP_400_BAD_REQUEST)
+
+    return_object = {'target_rel_path': tgt_path}
+
+    return HttpResponse(
+        json.dumps(return_object),
+        content_type='application/json'
+    )
+
+
+def is_folder(istorage, path):
+    """ return True if the thing is a folder """
+    dir, base = os.path.split(path)
+    try:
+        contents = istorage.listdir(dir)
+        if base in contents[0]:  # folders
+            return True
+        else:
+            return False
+    except SessionException:
+        return False
```

Comment:
Duplicate function with ```irods_path_is_directory()``` - only one is needed and the other duplicate one needs to be removed. I think this function needs to accommodate the case that input path contains a trailing slash.